File: server-go/crypto.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
)

// AES-256-GCM payload encryption, matching the C++ EncryptionService wire format
// ({encrypted, iv, authTag} base64). Enabled when QUEEN_ENCRYPTION_KEY (64 hex
// chars = 32 bytes) is set. This adds real per-message crypto CPU to push (seal)
// and pop (open) — exactly the feature cost we want to measure.
type Crypto struct {
	gcm cipher.AEAD
}

func NewCrypto(hexKey string) *Crypto {
	if len(hexKey) != 64 {
		return nil
	}
	key, err := hex.DecodeString(hexKey)
	if err != nil || len(key) != 32 {
		return nil
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil
	}
	g, err := cipher.NewGCM(block)
	if err != nil {
		return nil
	}
	return &Crypto{gcm: g}
}

// encryptPayload seals plaintext and returns the {encrypted,iv,authTag} object
// bytes (base64 fields), as the C++ push path stores it.
func (c *Crypto) encryptPayload(plaintext []byte) []byte {
	iv := make([]byte, c.gcm.NonceSize())
	_, _ = rand.Read(iv)
	sealed := c.gcm.Seal(nil, iv, plaintext, nil)
	// GCM appends the 16-byte tag; split it out to match the C++ shape.
	tag := sealed[len(sealed)-16:]
	ct := sealed[:len(sealed)-16]
	obj := map[string]string{
		"encrypted": base64.StdEncoding.EncodeToString(ct),
		"iv":        base64.StdEncoding.EncodeToString(iv),
		"authTag":   base64.StdEncoding.EncodeToString(tag),
	}
	b, _ := json.Marshal(obj)
	return b
}

// decryptField opens a base64 {encrypted,iv,authTag} back to plaintext bytes.
func (c *Crypto) decrypt(encB64, ivB64, tagB64 string) ([]byte, bool) {
	ct, e1 := base64.StdEncoding.DecodeString(encB64)
	iv, e2 := base64.StdEncoding.DecodeString(ivB64)
	tag, e3 := base64.StdEncoding.DecodeString(tagB64)
	if e1 != nil || e2 != nil || e3 != nil {
		return nil, false
	}
	sealed := append(ct, tag...)
	pt, err := c.gcm.Open(nil, iv, sealed, nil)
	if err != nil {
		return nil, false
	}
	return pt, true
}

File: server-rust/src/crypto.rs
use aes_gcm::aead::generic_array::GenericArray;
use aes_gcm::aead::{Aead, KeyInit, OsRng};
use aes_gcm::{AeadCore, Aes256Gcm, Key};
use base64::{engine::general_purpose::STANDARD as B64, Engine};

/// AES-256-GCM payload encryption (C++ EncryptionService wire format:
/// {encrypted, iv, authTag} base64). Real per-message crypto CPU.
pub struct Crypto {
    cipher: Aes256Gcm,
}

impl Crypto {
    pub fn new(hex_key: &str) -> Option<Crypto> {
        if hex_key.len() != 64 {
            return None;
        }
        let key_bytes = hex::decode(hex_key).ok()?;
        let key = Key::<Aes256Gcm>::from_slice(&key_bytes);
        Some(Crypto { cipher: Aes256Gcm::new(key) })
    }

    /// Returns the {encrypted,iv,authTag} object bytes for a plaintext payload.
    pub fn encrypt_payload(&self, plaintext: &[u8]) -> Vec<u8> {
        let nonce = Aes256Gcm::generate_nonce(&mut OsRng); // 12 bytes
        let sealed = self.cipher.encrypt(&nonce, plaintext).unwrap_or_default();
        let (ct, tag) = if sealed.len() >= 16 {
            sealed.split_at(sealed.len() - 16)
        } else {
            (&sealed[..], &sealed[..])
        };
        format!(
            "{{\"encrypted\":\"{}\",\"iv\":\"{}\",\"authTag\":\"{}\"}}",
            B64.encode(ct),
            B64.encode(nonce.as_slice()),
            B64.encode(tag)
        )
        .into_bytes()
    }

    pub fn decrypt(&self, enc_b64: &str, iv_b64: &str, tag_b64: &str) -> Option<Vec<u8>> {
        let ct = B64.decode(enc_b64).ok()?;
        let iv = B64.decode(iv_b64).ok()?;
        let tag = B64.decode(tag_b64).ok()?;
        let mut sealed = ct;
        sealed.extend_from_slice(&tag);
        let nonce = GenericArray::from_slice(&iv);
        self.cipher.decrypt(nonce, sealed.as_slice()).ok()
    }
}
